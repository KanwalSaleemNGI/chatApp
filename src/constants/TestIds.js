export default {
  authButton: {
    title: 'title',
    onPress: 'onPress',
  },
  chatInfo: {
    onPress: 'onPress',
  },
  login: {
    emailInput: 'emailInput',
    passwordInput: 'passwordInput',
  },
};
