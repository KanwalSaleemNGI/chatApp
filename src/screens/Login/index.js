import React, {useState, useRef, useEffect} from 'react';
import {
  View,
  Text,
  TouchableWithoutFeedback,
  Keyboard,
  Image,
  TouchableOpacity,
  ScrollView,
  Alert,
  Button,
  SafeAreaView,
  TextInput,
  KeyboardAvoidingView,
  Platform,
} from 'react-native';
import {useForm} from 'react-hook-form';
import {AuthButton, ShowLoader, Input} from '../../components';
import {
  loginHandler,
  googleSignInHandler,
  fbSignInHandler,
} from '../../store/actionCreators/auth/auth';
import {useDispatch, useSelector} from 'react-redux';
import {GoogleSignin} from '@react-native-google-signin/google-signin';
import styles from './style';
import Colors from '../../constants/Colors';

const LoginForm = ({navigation}) => {
  const dispatch = useDispatch();
  const isLoading = useSelector(state => state.auth.isLoading);

  const emailRef = useRef();
  const passwordRef = useRef();

  const {
    control,
    handleSubmit,
    formState: {errors},
    register,
    reset,
    setError,
    // setValue,
  } = useForm();

  const email = register('email');
  const password = register('password');

  const onSubmit = data => {
    console.log(data);
    dispatch(loginHandler(data));
  };

  const googleSignIn = () => {
    dispatch(googleSignInHandler());
  };

  const facebookSignIn = () => {
    dispatch(fbSignInHandler());
  };

  const [inputValue, setInputValue] = useState('');
  console.log(errors);
  return isLoading ? (
    <ShowLoader />
  ) : (
    <KeyboardAvoidingView
      behavior={Platform.OS === 'ios' ? 'padding' : 'height'}
      style={styles.container}>
      <TouchableWithoutFeedback
        onPress={() => Keyboard.dismiss()}
        testID="loginView">
        <View style={styles.screen}>
          <Input
            testID="emailInput"
            style={styles.input}
            container={{marginTop: 60}}
            control={control}
            rules={{
              required: 'Email is Required',
              pattern: {
                value: /^\w+([.-]?\w+)*@\w+([.-]?\w+)*(\.\w{2,3})+$/,
                message: 'Must be a valid Email',
              },
            }}
            errors={errors.email}
            name="email"
            placeholder="Email"
            ref={e => {
              email.ref(e);
              emailRef.current = e;
            }}
            onSubmitEditing={() => {
              passwordRef.current.focus();
            }}
            blurOnSubmit={false}
            returnKeyType="next"
            keyboardType="email-address"
          />
          {errors.email && (
            <Text style={styles.errorText} allowFontScaling={false}>
              {errors.email.message}
            </Text>
          )}
          <Input
            testID="passwordInput"
            style={styles.input}
            control={control}
            name="password"
            placeholder="Password"
            rules={{
              required: 'Password is Required',
            }}
            errors={errors.password}
            ref={e => {
              password.ref(e);
              passwordRef.current = e;
            }}
            onSubmitEditing={() => {
              passwordRef.current.blur();
            }}
            secureTextEntry={true}
            blurOnSubmit={false}
            returnKeyType="next"
          />
          {errors.password && (
            <Text style={styles.errorText} allowFontScaling={false}>
              {errors.password.message}
            </Text>
          )}

          <AuthButton
            style={styles.buttonContainer}
            onPress={handleSubmit(onSubmit)}
            testID="loginButton">
            Login
          </AuthButton>

          <View style={styles.conditionContainer}>
            <Text style={styles.text} allowFontScaling={false}>
              Donâ€™t have an account ?
            </Text>
            <TouchableOpacity
              testID="signUpButton"
              activeOpacity={0.7}
              onPress={() => navigation.navigate('signUpForm')}>
              <Text
                style={[styles.text, {color: Colors.primary}]}
                allowFontScaling={false}>
                Sign up
              </Text>
            </TouchableOpacity>
          </View>

          <View style={styles.socialMainContainer}>
            <Text style={styles.text} allowFontScaling={false}>
              Login with
            </Text>
            <View style={styles.socialContainer}>
              <TouchableOpacity
                style={styles.socialLogoContainer}
                onPress={googleSignIn}>
                <Image
                  source={require('../.././assets/images/googleImg.jpg')}
                  style={styles.socialLogo}
                />
              </TouchableOpacity>
              <TouchableOpacity
                style={styles.socialLogoContainer}
                onPress={facebookSignIn}>
                <Image
                  source={require('../.././assets/images/facebookImg.jpg')}
                  style={styles.socialLogo}
                />
              </TouchableOpacity>
            </View>
          </View>
        </View>
      </TouchableWithoutFeedback>
    </KeyboardAvoidingView>
  );
};

export default LoginForm;
