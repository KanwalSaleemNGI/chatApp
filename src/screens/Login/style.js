import {StyleSheet} from 'react-native';
import Colors from '../../constants/Colors';

const styles = StyleSheet.create({
  screen: {
    flexGrow: 1,
    backgroundColor: Colors.backgroundColor,
    alignItems: 'center',
    paddingTop: 20,
  },
  errorText: {
    alignSelf: 'flex-start',
    marginLeft: 20,
    color: Colors.common.errorColor,
    fontFamily: 'Roboto-Regular',
  },
  buttonContainer: {
    marginTop: 30,
  },
  conditionContainer: {
    flexDirection: 'row',
    marginVertical: 20,
  },
  text: {
    fontSize: 16,
    color: Colors.black,
    fontFamily: 'Roboto-Bold',
    marginHorizontal: 2,
  },
  socialMainContainer: {
    alignItems: 'center',
    width: '100%',
    marginTop: 20,
  },
  socialContainer: {
    marginTop: 10,
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'center',
  },
  socialLogoContainer: {
    marginHorizontal: 10,
  },
  socialLogo: {
    width: 40,
    height: 40,
  },
});

export default styles;
