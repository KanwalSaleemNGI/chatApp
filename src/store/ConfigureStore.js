import React from 'react';
import {createStore, combineReducers, applyMiddleware} from 'redux';
import ReduxThunk from 'redux-thunk';
import authReducer from './reducers/auth';
import dashboardReducer from './reducers/dashboard';
import {composeWithDevTools} from 'redux-devtools-extension';
import {persistStore, persistReducer, createTransform} from 'redux-persist';
import AsyncStorage from '@react-native-async-storage/async-storage';
import {reducer as network} from 'react-native-offline';
import {createNetworkMiddleware} from 'react-native-offline';
import {sendMessageAsync} from './actionCreators/dashboard/chat';

// We have to map our actions to an object
const actions = {
  sendMessageAsync,
};

// Transform how the persistor reads the network state
const networkTransform = createTransform(
  (inboundState, key) => {
    const actionQueue = [];

    inboundState.actionQueue.forEach(action => {
      if (typeof action === 'function') {
        actionQueue.push({
          function: action.meta.name,
          args: action.meta.args,
        });
      } else if (typeof action === 'object') {
        actionQueue.push(action);
      }
    });

    return {
      ...inboundState,
      actionQueue,
    };
  },
  (outboundState, key) => {
    const actionQueue = [];

    outboundState.actionQueue.forEach(action => {
      if (action.function) {
        const actionFunction = actions[action.function];
        actionQueue.push(actionFunction(...action.args));
      } else {
        actionQueue.push(action);
      }
    });

    return {...outboundState, actionQueue};
  },
  // The 'network' key may change depending on what you
  // named your network reducer.
  {whitelist: ['network']},
);

const persistConfig = {
  key: 'root',
  storage: AsyncStorage,
  // blacklist: ['chat']  => the name of the blacklist is the key of the reducer If you donâ€™t want to persist a part of your state you could put it in the blacklist.
  //whitelist: ['auth'] //The whitelist is set up in the same way as the blacklist except that it defines the parts of state that you do want to persist.
  blacklist: ['auth', 'dashboard'],
  whitelist: ['network'],
  transforms: [networkTransform], // Add the transform into the persist config
};

const authPersistConfig = {
  key: 'auth',
  storage: AsyncStorage,
  whitelist: ['userDetails'],
  blacklist: ['isLoading'],
};

const dashboardPersistConfig = {
  key: 'dashboard',
  storage: AsyncStorage,
  whitelist: ['allChats', 'allUsers'],
};

const rootReducer = combineReducers({
  auth: persistReducer(authPersistConfig, authReducer),
  dashboard: persistReducer(dashboardPersistConfig, dashboardReducer),
  network,
});

const persistedReducer = persistReducer(persistConfig, rootReducer);
const networkMiddleware = createNetworkMiddleware({
  queueReleaseThrottle: 200,
});

export const store = createStore(
  persistedReducer,
  composeWithDevTools(applyMiddleware(networkMiddleware, ReduxThunk)),
);

export const persistor = persistStore(store);
