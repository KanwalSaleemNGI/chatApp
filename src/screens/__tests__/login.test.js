import React from 'react';
import Login from '../Login/index';
import {fireEvent, render, act} from '@testing-library/react-native';
import {Provider} from 'react-redux';
import {store} from '../../store/ConfigureStore';
import TestIds from '../../constants/TestIds';

// shows email is required
// shows password is required
// shows invalid email
//handled valid input submission

describe('Login Comp', () => {
  const emailIsRequired = 'Email is Required';
  const passwordIsRequired = 'Password is Required';
  const mustBeAValidEmail = 'Must be a valid Email';
  // const navigationMock = jest.fn();
  const renderHandler = () => {
    return render(
      <Provider store={store}>
        {/* <Login navigation={{navigate: navigationMock}} /> */}
        <Login />
      </Provider>,
    );
  };

  //snapshot match test
  it('it should matches to snapshot', () => {
    const {toJSON} = renderHandler();

    expect(toJSON()).toMatchSnapshot();
  });

  //rendered elements test
  it('it should renders default elements', () => {
    const {getByPlaceholderText, getByText} = renderHandler();

    getByText('Login');

    getByPlaceholderText('Email');
    getByPlaceholderText('Password');
  });

  //client side validation
  it('it should shows error message if all fields are empty', async () => {
    const {getByTestId, getByText} = renderHandler();

    await fireEvent.press(getByTestId(TestIds.authButton.onPress));
    getByText(emailIsRequired);
    getByText(passwordIsRequired);
  });

  it('it should shows "Email is Required" error message', async () => {
    const {getByTestId, getByText, queryAllByText} = renderHandler();

    fireEvent.changeText(getByTestId(TestIds.login.passwordInput), '123456');
    await fireEvent.press(getByTestId(TestIds.authButton.onPress));
    getByText(emailIsRequired);
    expect(queryAllByText(passwordIsRequired).length).toBe(0);

    // debug();
  });

  it('it should shows "Password is Required" error message', async () => {
    const {getByTestId, getByText, queryAllByText} = renderHandler();

    fireEvent.changeText(
      getByTestId(TestIds.login.emailInput),
      'aaa@gmail.com',
    );
    await fireEvent.press(getByTestId(TestIds.authButton.onPress));
    getByText(passwordIsRequired);
    expect(queryAllByText(emailIsRequired).length).toBe(0);

    // debug();
  });

  it('it should shows "Must be a valid Email" error message', async () => {
    const {getByTestId, getByText, queryAllByText} = renderHandler();

    fireEvent.changeText(getByTestId(TestIds.login.emailInput), 'aaa');
    fireEvent.changeText(getByTestId(TestIds.login.passwordInput), '123456');
    await fireEvent.press(getByTestId(TestIds.authButton.onPress));
    getByText(mustBeAValidEmail);

    expect(queryAllByText(emailIsRequired).length).toBe(0);
    expect(queryAllByText(passwordIsRequired).length).toBe(0);

    // debug();
  });

  it('it should handles valid input submission', async () => {
    // fetch.mockResponseOnce(JSON.stringify({passes: true}));
    const {getByTestId} = renderHandler();

    fireEvent.changeText(
      getByTestId(TestIds.login.emailInput),
      'aaa@gmail.com',
    );
    fireEvent.changeText(getByTestId(TestIds.login.passwordInput), '123456');
    await fireEvent.press(getByTestId(TestIds.authButton.onPress));
  });
});
