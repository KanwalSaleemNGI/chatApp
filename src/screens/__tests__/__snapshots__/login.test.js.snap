// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Comp it should matches to snapshot 1`] = `
<View
  onLayout={[Function]}
  style={
    Array [
      Object {
        "backgroundColor": "#E5E5E5",
        "flexGrow": 1,
      },
      Object {
        "paddingBottom": 0,
      },
    ]
  }
>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "paddingTop": 20,
      }
    }
    testID="loginView"
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#F6F6F6",
            "borderColor": "#E8E8E8",
            "borderRadius": 10,
            "borderWidth": 1,
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginVertical": 10,
            "width": "90%",
          },
          undefined,
          undefined,
        ]
      }
    >
      <TextInput
        blurOnSubmit={false}
        focusable={true}
        keyboardType="email-address"
        onChangeText={[Function]}
        onSubmitEditing={[Function]}
        placeholder="Email"
        placeholderTextColor="#000000"
        returnKeyType="next"
        style={
          Array [
            Object {
              "color": "#000000",
              "flexBasis": "100%",
              "fontFamily": "Roboto-Regular",
              "fontSize": 16,
              "height": 50,
              "paddingHorizontal": 10,
            },
            undefined,
          ]
        }
        testID="emailInput"
        value=""
      />
    </View>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#F6F6F6",
            "borderColor": "#E8E8E8",
            "borderRadius": 10,
            "borderWidth": 1,
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginVertical": 10,
            "width": "90%",
          },
          undefined,
          undefined,
        ]
      }
    >
      <TextInput
        blurOnSubmit={false}
        focusable={true}
        onChangeText={[Function]}
        onSubmitEditing={[Function]}
        placeholder="Password"
        placeholderTextColor="#000000"
        returnKeyType="next"
        secureTextEntry={true}
        style={
          Array [
            Object {
              "color": "#000000",
              "flexBasis": "100%",
              "fontFamily": "Roboto-Regular",
              "fontSize": 16,
              "height": 50,
              "paddingHorizontal": 10,
            },
            undefined,
          ]
        }
        testID="passwordInput"
        value=""
      />
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#4C96D6",
          "borderRadius": 25,
          "elevation": 5,
          "height": 50,
          "justifyContent": "center",
          "marginTop": 30,
          "opacity": 1,
          "shadowColor": "black",
          "shadowOffset": Object {
            "height": 2,
            "width": 0,
          },
          "shadowOpacity": 0.26,
          "shadowRadius": 8,
          "width": "90%",
        }
      }
      testID="onPress"
    >
      <Text
        allowFontScaling={false}
        style={
          Object {
            "color": "#fff",
            "fontFamily": "Roboto-Regular",
            "fontSize": 16,
          }
        }
        testID="title"
      >
        Login
      </Text>
    </View>
    <View
      style={
        Object {
          "flexDirection": "row",
          "marginVertical": 20,
        }
      }
    >
      <Text
        allowFontScaling={false}
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Roboto-Bold",
            "fontSize": 16,
            "marginHorizontal": 2,
          }
        }
      >
        Donâ€™t have an account ?
      </Text>
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
        testID="signUpButton"
      >
        <Text
          allowFontScaling={false}
          style={
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "Roboto-Bold",
                "fontSize": 16,
                "marginHorizontal": 2,
              },
              Object {
                "color": "#4C96D6",
              },
            ]
          }
        >
          Sign up
        </Text>
      </View>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "marginTop": 20,
          "width": "100%",
        }
      }
    >
      <Text
        allowFontScaling={false}
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Roboto-Bold",
            "fontSize": 16,
            "marginHorizontal": 2,
          }
        }
      >
        Login with
      </Text>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "center",
            "marginTop": 10,
          }
        }
      >
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "marginHorizontal": 10,
              "opacity": 1,
            }
          }
        >
          <Image
            source={
              Object {
                "testUri": "../../../src/assets/images/googleImg.jpg",
              }
            }
            style={
              Object {
                "height": 40,
                "width": 40,
              }
            }
          />
        </View>
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "marginHorizontal": 10,
              "opacity": 1,
            }
          }
        >
          <Image
            source={
              Object {
                "testUri": "../../../src/assets/images/facebookImg.jpg",
              }
            }
            style={
              Object {
                "height": 40,
                "width": 40,
              }
            }
          />
        </View>
      </View>
    </View>
  </View>
</View>
`;
